lib/active_model/serializer/type.rb

```diff
@@ -16,8 +16,24 @@ module ClassMethods
         # @example
         #   class AdminAuthorSerializer < ActiveModel::Serializer
         #     type 'authors'
-        def type(type)
-          self._type = type && type.to_s
+        #
+        # @example
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type { |object| object.class.name }
+        #
+        # @example
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type proc { |object| object.class.name }
+        #
+        def type(type = nil, &block)
+          type = block if block_given?
+          return unless type
+
+          self._type = if type.respond_to?(:call) && type.arity == 1
+                         type
+                       else
+                         type.to_s
```

Comment:
You could put it [there](https://github.com/rails-api/active_model_serializers/pull/1777/files#diff-65abd6a4fd7fd55fe487ff138e0841dbR24). Thus, this method becomes:
```ruby
def type(type = nil, &block)
  self._type = block || type
end